spec.md

```diff
@@ -0,0 +1,461 @@
+
+### 1.1. Conventions in this Document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+The grammatical rules in this document are to be interpreted as described in [RFC5234](https://datatracker.ietf.org/doc/html/rfc5234). We are using the following core rules:
+
+```abnf
+CR    = %x0D     ; carriage return
+LF    = %x0A     ; line feed
+CRLF  = CR LF
+DIGIT = %x30-39  ; 0-9
+HTAB  = %x09     ; horizontal tab
+SP    = %x20     ; space
+```
+
+
+
+## 2. General Structure
+
+UltraStar files are plain text files. The UTF-8 encoding MUST be used. Implementations MUST NOT add a byte order mark to the beginning of a file. In the interests of interoperability, implementations MAY ignore the presence of a byte order mark rather than treating it as an error.
+
+> [!CAUTION]
+>
+> Whether files with a BOM may be accepted is an open discussion (#46).
+
+The canonical file extension for the UltraStar file format is `.txt`.
+
+UltraStar files consist of a header and a body. The header contains metadata about the song and the file. The body contains the musical data. A file SHOULD end with an `E` on a single line. Everything After a trailing `E` MUST be ignored.
+
+> [!CAUTION]
+>
+> Whether the trailing `E` is optional or not is open for discussion (#44).
+
+```abnf
+UltraStar-file = file-header
+                 file-body
+                 [ %x45 *char ]  ; E
+char           = %x00-10FFFF
+```
+
+### 2.1. Line Endings and White Space
+
+Both the header and the body of a file are defined in terms of lines. A line is a string of text that is terminated with an end-of-line sequence.
+
+```abnf
+end-of-line = ( CR / LF / CRLF )
+empty-line  = end-of-line
+```
+
+Implementations SHOULD use a single Line Feed (`%x0A`) as line terminator. Implementations MUST accept the end of input (`EOF`) as a valid line terminator. Empty lines are ignored throughout the entire file.
+
+> [!CAUTION]
+>
+> The allowed end-of-line sequences as well as potential recommendations for implementations are currently open for discussion (#43).
+
+> [!CAUTION]
+>
+> Whether empty lines are allowed or not is currently open for discussion (#43). Whether a line that consists only of whitespace is recognized as an empty line is not yet decided.
+
+Whitespace is used as a separator in many places of the format. Only space (`%x20`) and horizontal tab (`%x09`) are recognized as whitespace characters. In particular other unicode characters with the property `White_Space=yes` MUST NOT be treated as whitespace characters in the context of this specification.
+
+```abnf
+WSP = ( SP / HTAB )
+```
+
+> [!CAUTION]
+>
+> Definition and handling of whitespace characters is currently open for discussion (#46).
```

Comment:
Might be good to add a link to the issue for now. Then when the PR is ready to be merged we should clean up these sections as best we can.